File: src/main/java/uk/co/reecedunn/intellij/plugin/xquery/resolve/reference/XQueryVariableNameReference.java
package uk.co.reecedunn.intellij.plugin.xquery.resolve.reference;

import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiReferenceBase;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import uk.co.reecedunn.intellij.plugin.core.functional.Option;
import uk.co.reecedunn.intellij.plugin.xquery.ast.xquery.XQueryEQName;
import uk.co.reecedunn.intellij.plugin.xquery.psi.XQueryVariable;
import uk.co.reecedunn.intellij.plugin.xquery.psi.XQueryVariableResolver;

public class XQueryVariableNameReference extends PsiReferenceBase<XQueryEQName> {
    public XQueryVariableNameReference(XQueryEQName element, TextRange range) {
        super(element, range);
    }

    @Nullable
    @Override
    public PsiElement resolve() {
        XQueryEQName name = getElement();

        PsiElement element = name;
        while (element != null) {
            if (element instanceof XQueryVariableResolver) {
                Option<XQueryVariable> resolved = ((XQueryVariableResolver)element).resolveVariable(name);
                if (resolved.isDefined()) {
                    return resolved.get().getVariable();
                }
            }

            PsiElement next = element.getPrevSibling();
            if (next == null) {
                next = element.getParent();
            }
            element = next;
        }
        return null;
    }

    @NotNull
    @Override
    public Object[] getVariants() {
        return new Object[0];
    }
}

File: src/main/java/uk/co/reecedunn/intellij/plugin/xquery/resolve/reference/XQueryVariableNameReference.kt
/*
 * Copyright (C) 2016-2018 Reece H. Dunn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.reecedunn.intellij.plugin.xquery.resolve.reference

import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiReferenceBase
import uk.co.reecedunn.intellij.plugin.xpath.ast.xpath.XPathEQName
import uk.co.reecedunn.intellij.plugin.xpath.model.inScopeVariablesForFile

class XQueryVariableNameReference(element: XPathEQName, range: TextRange) : PsiReferenceBase<XPathEQName>(element, range) {
    override fun resolve(): PsiElement? {
        val name = element
        val match = name.inScopeVariablesForFile().find { variable ->
            val qname = variable.variableName!!
            val matchPrefix = name.prefix?.text == (qname.prefix?.staticValue as? String)
            val matchLocalName = name.localName?.text == (qname.localName?.staticValue as? String)
            matchPrefix && matchLocalName
        }
        return match?.variableName?.declaration?.get() as? PsiElement
    }

    override fun getVariants(): Array<Any> {
        return arrayOf()
    }
}
